docs/lab/lab04/lab04.md

```diff
@@ -0,0 +1,407 @@
+
+```rust
+// Initialize ADC pin
+let mut adc_pin = Channel::new_pin(peripherals.PIN_X, Pull::None); // where X should be replaced with a pin number
+```
+
+Once we have the ADC and pin set up, we can start reading values from the pin. 
+
+```rust
+let level = adc.read(&mut adc_pin).await.unwrap(); // read a value from the pin
+info!("Light sensor reading: {}", level); // print the value over serial
+Timer::after_secs(1).await; // wait a bit before reading and printing another value
+```
+
+## Exercises (not final version)
+
+1. Connect an LED to pin GP2 and a photo-resistor to ADC0. Use [KiCad](https://www.kicad.org/) to draw the schematics. (**1p**)
+2. Modify the provided working example (`lab04_ex2`) to light the led of your circuit to 25% intensity. (**1p**)
+3. Increase the led's intensity every second, until it reaches max intensity, when it stops. (**1p**)
+3. Read the value of the photo-resistor and print it to the console. (**2p**)
+
+:::info
+To see the console with messages from the Pico, use the flash command with an extra `-s` parameter. 
+Example:
+```bash
+elf2uf2-rs -s -d /target/thumbv6m-none-eabi/debug/<crate_name>
+```
+:::
+
+:::info
+To be able to print messages to the console, we need to send messages over a simulated serial port to the computer. For this, we will use the USB driver provided by Embassy.
+
+```rust
+use embassy_rp::usb::{Driver, InterruptHandler};
+use embassy_rp::{bind_interrupts, peripherals::USB};
+use log::info;
+
+// Use for the serial over USB driver
+bind_interrupts!(struct Irqs {
+    USBCTRL_IRQ => InterruptHandler<USB>;
+});
+
+
+// The task used by the serial port driver 
+// over USB
+#[embassy_executor::task]
+async fn logger_task(driver: Driver<'static, USB>) {
+    embassy_usb_logger::run!(1024, log::LevelFilter::Info, driver);
+}
+
+#[embassy_executor::main]
+async fn main(spawner: Spawner) {
+    let peripherals = embassy_rp::init(Default::default());
+
+    // Start the serial port over USB driver
+    let driver = Driver::new(peripherals.USB, Irqs);
+    spawner.spawn(logger_task(driver)).unwrap();
+
+    // ...
+
+    info!("message");
+}
+
+```
+:::
+
+:::warning
+Notice that the USB driver also uses an `InterruptHandler` import that could be confused with the `InterruptHandler` used by ADC. Make sure to use different naming conventions for each one, as described in the warning [here](#adc-in-embassy-rs). 
+:::
+4. Depending on the value read from the photo-resistor, brighten or dim the led. The led should shine brighter when there is *less* light in the room. (**2p**)
+:::tip
+Use the serial console to debug your program!
+:::
+5. Remove the photoresistor and the led from the circuit. Instead, connect a button to GP0 and an RGB led to pins GP1, GP2 and GP3. Make the RGB switch from red -> yellow -> green every time the button is pressed. (**2p**)
+6. Using the `SysTick` interrupt in *bare metal*, make the led blink at a 100ms delay. (**1p**)
```

Comment:
I would add a hint that they need to blink the LED in bare metal, link the code form the previous lab and that an `AtomicBool` is required of the shared state.

Reply:
Noted. I will solve this exercise of the lab today and provide more tips for the students.